core/kazoo_number_manager/src/carriers/knm_voxbone.erl

```diff
@@ -0,0 +1,332 @@
+generate_number_summary(#{id := _CartId, query_id := QueryId}=Cart, [Order | Rest], Acc) ->
+    case kz_json:get_value(<<"productType">>, Order) =:= <<"DID">> of
+        'false' ->
+            generate_number_summary(Cart, Rest, Acc);
+        'true' ->
+            Options = [{<<"orderReference">>, kz_json:get_value(<<"orderReference">>, Order)}
+                      ] ++ knm_voxbone_util:required_params(),
+            {'ok', DIDs} = knm_voxbone_util:voxbone_request('get', <<"inventory/did">>, Options),
+            KNMs = order_to_KNM(QueryId, DIDs),
+            generate_number_summary(Cart, Rest, KNMs ++ Acc)
+    end.
+
+%%------------------------------------------------------------------------------
+%% @doc convert voxbone dids into knm phone numbers
+%% @end
+%%------------------------------------------------------------------------------
+-spec order_to_KNM(kz_term:ne_binary(), kz_json:objects()) -> [{knm_search:option(), {kz_term:ne_binary(), kz_term:atom(), kz_term:atom(), kz_json:object()}}].
+order_to_KNM(QueryId, DIDs) ->
+    JObjs = kz_json:get_value(<<"dids">>, DIDs),
+    order_to_KNM(QueryId, JObjs, []).
+
+-spec order_to_KNM(kz_term:ne_binary(), kz_json:objects(), knm_number:phone_numbers()) -> [{knm_search:option(), {kz_term:ne_binary(), kz_term:atom(), kz_term:atom(), kz_json:object()}}].
+order_to_KNM(_QueryId, [], Acc) ->
+    Acc;
+order_to_KNM(QueryId, [DID | Rest], Acc) ->
+    MasterAccountId = kapps_config:get(<<"accounts">>, <<"master_account_id">>),
+    Options = [{'auth_by', MasterAccountId}
+              ,{'assign_to', MasterAccountId}
+              ,{'state', ?NUMBER_STATE_AVAILABLE}
+              ,{'carrier_data', DID}
+              ,{'module_name', <<"knm_voxbone">>}],
+    DID2 = kz_json:get_value(<<"e164">>, DID),
+    Number = {DID2, ?MODULE, ?NUMBER_STATE_AVAILABLE, DID},
+    %% Add number to database to ensure inventory is updated before acquire_number stub
+    {'ok', _} = knm_number:create(DID2, Options),
+    order_to_KNM(QueryId, Rest, [{QueryId, Number} | Acc]).
+
+%%------------------------------------------------------------------------------
+%% @doc cleanup voxbone cart artifact when order can't be satisfied.
+%% @end
+%%------------------------------------------------------------------------------
+-spec maybe_destroy_cart(map()) -> {'ok', kz_json:object()} | {'error', any()}.
+maybe_destroy_cart(#{id := CartId, query_id := _QueryId}=_Cart) ->
+    knm_voxbone_util:voxbone_request('delete', <<"ordering/cart/",(kz_term:to_binary(CartId))/binary>>, [], []).
+
+%%------------------------------------------------------------------------------
+%% @doc fetches DID metadata from voxbone portal
+%% @end
+%%------------------------------------------------------------------------------
+-spec fetch_did(knm_number:phone_number()) -> {'ok', kz_json:object()} | {'error', any()}.
+fetch_did(Number) ->
+    Options = kz_json:from_list([{<<"e164pattern">>, knm_voxbone_util:to_voxbone_pattern(Number)}
```

Comment:
Wouldn't `kz_json:from_list()` return `{[...]}` and crash when appended to `required_params()` return?